Sources/Articles/Documentation.docc/2025/gha-faster-ci/gha-faster-ci.md

```diff
@@ -0,0 +1,448 @@
+on:
+  push: { branches: [main] }
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Build image
+        run: docker build --network=host -t app:latest .
+
+      # Push the image to a Docker container registry and deploy the app
+```
+
+These CIs usually take 10 minutes in tests and 14 minutes 30 seconds for deployments in [Penny](https://github.com/vapor/penny-bot).
+That's too much time wasted waiting. How can you improve these CI times?
+
+![Tests CI Initial State](tests-ci-initial.png)
+![Deployment CI Initial State](deploy-ci-initial.png)
+
+## Use caching
+
+Luckily for you, GitHub provides an [official caching action](https://github.com/actions/cache) that you can leverage in our CI.
+For this, you need to make sure you cache SwiftPM's `.build` directory after your builds have succeeded, so the next build can restore those build artifacts for faster build times.
+It'll look like this in the tests CI:
+
+```diff
++      - name: Restore .build
++        id: "restore-build"
++        uses: actions/cache/restore@v4
++        with:
++          path: .build
++          key: "swiftpm-tests-build-${{ runner.os }}-${{ github.event.pull_request.base.sha || github.event.after }}"
++          restore-keys: "swiftpm-tests-build-${{ runner.os }}-"
+
+      - name: Run unit tests
+        run: swift test --enable-code-coverage
+
++      - name: Cache .build
++        if: steps.restore-build.outputs.cache-hit != 'true'
++        uses: actions/cache/save@v4
++        with:
++          path: .build
++          key: "swiftpm-tests-build-${{ runner.os }}-${{ github.event.pull_request.base.sha || github.event.after }}"
+```
+
+In the "Restore .build" step, you're using:
+* The `restore` capability of `actions/cache`, and asking it to restore your previously-uploaded cache to the `.build` path.
+* The `key` as a way to uniquely identify different caches. This resolves to a string like `swiftpm-tests-build-Linux-f5ded47aafafe1f9f542e833a5f3dc01970bbaee`.
+* The `restore-keys` as fallback. The action will restore the latest cache that starts with `restore-keys`, if no cache matching `key` was found.
+
+Note that `actions/cache` follows some branch protection rules so it's not vulnerable to cache poisoning attacks.
+That means that it'll only restore a cache if it was cached from the current branch or the primary branch of your repository.
+
+After the tests process is over, you can cache the build artifacts to make sure the next CI run can leverage this fresh cache.
+
+So in the "Cache .build" step, you're caching your `.build` directory with the same key as the one in the "Restore .build" step.
+
+With the `if` condition, you're avoiding spending time in the cache step when you already found an **exact match** for the cache key in the "Restore .build" step. This is because `actions/cache` rejects cache entries with duplicate names, and does not have an update mechanism.
```

Comment:
I'm not sure this is really needed? Couldn't you just use `actions/cache@v4` in the first step and depend on the default behaviour? I believe it works like you would expect (i.e. save the new `.build` if an exact match hasn't been found)

Reply:
> I'm not sure this is really needed?

What exactly are you referring to by "this"?

By "first step" you mean "Restore .build"? what would it help to use "actions/cache" or "actions/cache/restore"? I assumed those do the same thing, while one being more explicit.